lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+{
+    return getVector3d<GfVec3f>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate,
+        UsdMayaXformStackTokens->scalePivotTranslate));
+}
+
+template<class V>
+Ufe::Vector3d UsdTransform3dMayaXformStack::getVector3d(
+    const TfToken& attrName
+) const
+{
+    // If the attribute doesn't exist yet, return a zero vector.
+    auto attr = prim().GetAttribute(attrName);
+    if (!attr) {
+        return Ufe::Vector3d(0, 0, 0);
+    }
+
+    UsdGeomXformOp op(attr);
+    TF_AXIOM(op);
+
+    V v;
+    op.Get(&v, getTime(path()));
+    return toUfe(v);
+}
+
+template<class V>
+Ufe::SetVector3dUndoableCommand::Ptr
+UsdTransform3dMayaXformStack::setVector3dCmd(
+    const V& v, const TfToken& attrName, const TfToken& opSuffix
+)
+{
+    UsdGeomXformOp op;
+
+    // If the attribute doesn't exist yet, create it.
+    auto attr = prim().GetAttribute(attrName);
+    if (!attr) {
+        InTransform3dChange guard(path());
+        op = _xformable.AddTranslateOp(OpPrecision<V>::precision, opSuffix);
+        if (!TF_VERIFY(op)) {
+            return nullptr;
+        }
+        op.Set(v);
+        setXformOpOrder();
+    }
+    else {
+        op = UsdGeomXformOp(attr);
+        TF_AXIOM(op);
+    }
+
+    return std::make_shared<UsdVecOpUndoableCmd<V>>(
+        v, path(), op, UsdTimeCode::Default());
+}
+
+Ufe::TranslateUndoableCommand::Ptr UsdTransform3dMayaXformStack::pivotCmd(
+    const TfToken& pvtOpSuffix,
+    double x, double y, double z
+)
+{
+    auto pvtAttrName = UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate, pvtOpSuffix);
+    UsdGeomXformOp p;
+
+    // If the pivot attribute pair doesn't exist yet, create it.
+    auto attr = prim().GetAttribute(pvtAttrName);
+    GfVec3f v(x, y, z);
+    if (!attr) {
+        // Without a notification guard each operation (each transform op
+        // addition, setting the attribute value, and setting the transform op
+        // order) will notify.  Observers would see an object in an inconsistent
+        // state, especially after pivot is added but before its inverse is
+        // added --- this does not match the Maya transform stack.  Use of
+        // SdfChangeBlock is discouraged when calling USD APIs above Sdf, so
+        // use our own guard.
+        InTransform3dChange guard(path());
+        p = _xformable.AddTranslateOp(UsdGeomXformOp::PrecisionFloat,
+                                     pvtOpSuffix);
+        auto pInv = _xformable.AddTranslateOp(UsdGeomXformOp::PrecisionFloat,
+            pvtOpSuffix, /* isInverseOp */ true);
+        if (!TF_VERIFY(p && pInv)) {
+            return nullptr;
+        }
+        p.Set(v);
+        setXformOpOrder();
+    }
+    else {
+        p = UsdGeomXformOp(attr);
+        TF_AXIOM(p);
+    }
+
+    return std::make_shared<UsdVecOpUndoableCmd<GfVec3f>>(
+        v, path(), p, UsdTimeCode::Default());
+}
+
+void UsdTransform3dMayaXformStack::setXformOpOrder()
```

Comment:
Transform op sorting method.  Since all transform ops have an index, just write them into a map using that index, then read back the map, which is by definition in order.